api/grpcserver/smesher_service.go

```diff
@@ -0,0 +1,155 @@
+	return &SmesherService{
+		Mining: miner,
+	}
+}
+
+// IsSmeshing reports whether the node is smeshing
+func (s SmesherService) IsSmeshing(ctx context.Context, in *empty.Empty) (*pb.IsSmeshingResponse, error) {
+	log.Info("GRPC SmesherService.IsSmeshing")
+
+	stat, _, _, _ := s.Mining.MiningStats()
+	isSmeshing := stat == activation.InitDone
+	return &pb.IsSmeshingResponse{IsSmeshing: isSmeshing}, nil
+}
+
+// StartSmeshing requests that the node begin smeshing
+func (s SmesherService) StartSmeshing(ctx context.Context, in *pb.StartSmeshingRequest) (*pb.StartSmeshingResponse, error) {
+	log.Info("GRPC SmesherService.StartSmeshing")
+
+	if in.Coinbase == nil {
+		return nil, status.Errorf(codes.InvalidArgument, "`Coinbase` must be provided")
+	}
+	if in.DataDir == "" {
+		return nil, status.Errorf(codes.InvalidArgument, "`DataDir` must be provided")
+	}
+	if in.CommitmentSize == nil {
+		return nil, status.Errorf(codes.InvalidArgument, "`CommitmentSize` must be provided")
+	}
+
+	addr := types.BytesToAddress(in.Coinbase.Address)
+	if err := s.Mining.StartPost(addr, in.DataDir, in.CommitmentSize.Value); err != nil {
+		log.Error("error starting post: %s", err)
+		return nil, status.Errorf(codes.Internal, "error initializing smeshing")
+	}
+	return &pb.StartSmeshingResponse{
+		Status: &rpcstatus.Status{Code: int32(code.Code_OK)},
+	}, nil
+}
+
+// StopSmeshing requests that the node stop smeshing
+func (s SmesherService) StopSmeshing(ctx context.Context, in *pb.StopSmeshingRequest) (*pb.StopSmeshingResponse, error) {
+	log.Info("GRPC SmesherService.StopSmeshing")
+
+	s.Mining.Stop()
+	return &pb.StopSmeshingResponse{
+		Status: &rpcstatus.Status{Code: int32(code.Code_OK)},
+	}, nil
+}
+
+// SmesherId returns the smesher ID of this node
+func (s SmesherService) SmesherId(ctx context.Context, in *empty.Empty) (*pb.SmesherIdResponse, error) {
+	log.Info("GRPC SmesherService.SmesherId")
+
+	smesherID := s.Mining.GetSmesherID()
+	return &pb.SmesherIdResponse{AccountId: &pb.AccountId{Address: smesherID.ToBytes()}}, nil
+}
+
+// Coinbase returns the current coinbase setting of this node
+func (s SmesherService) Coinbase(ctx context.Context, in *empty.Empty) (*pb.CoinbaseResponse, error) {
+	log.Info("GRPC SmesherService.Coinbase")
+
+	_, _, coinbase, _ := s.Mining.MiningStats()
+	addr, err := types.StringToAddress(coinbase)
+	if err != nil {
+		log.Error("error converting coinbase: %s", err)
+		return nil, status.Errorf(codes.Internal, "error reading coinbase data")
+	}
+	return &pb.CoinbaseResponse{AccountId: &pb.AccountId{Address: addr.Bytes()}}, nil
+}
+
+// SetCoinbase sets the current coinbase setting of this node
+func (s SmesherService) SetCoinbase(ctx context.Context, in *pb.SetCoinbaseRequest) (*pb.SetCoinbaseResponse, error) {
+	log.Info("GRPC SmesherService.SetCoinbase")
+
+	if in.Id == nil {
+		return nil, status.Errorf(codes.InvalidArgument, "`Id` must be provided")
+	}
+
+	addr := types.BytesToAddress(in.Id.Address)
```

Comment:
should we use the function that returns error? to make sure we don't receive invalid input, this function converts anything so it might be a bit dangerous

Reply:
Hmm. The code is dead simple. Do you think there's a potential issue here? I mean, literally any 20 bytes is a valid possible address. The only thing we could check is the length (which I think is checked implicitly in `SetBytes`)

```
func BytesToAddress(b []byte) Address {
	var a Address
	a.SetBytes(b)
	return a
}
```